src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,502 @@
+     * @dev Only callable by the paymentUpdater
+     */
+    function submitRoot(
+        bytes32 root,
+        uint32 paymentCalculationEndTimestamp
+    ) external onlyPaymentUpdater {
+        require(
+            paymentCalculationEndTimestamp > currPaymentCalculationEndTimestamp,
+            "PaymentCoordinator.submitRoot: new root must be for newer calculated period"
+        );
+        uint32 rootIndex = uint32(distributionRoots.length);
+        uint32 activatedAt = uint32(block.timestamp) + activationDelay;
+        distributionRoots.push(
+            DistributionRoot({
+                root: root,
+                activatedAt: activatedAt,
+                paymentCalculationEndTimestamp: paymentCalculationEndTimestamp
+            })
+        );
+        currPaymentCalculationEndTimestamp = paymentCalculationEndTimestamp;
+        emit DistributionRootSubmitted(rootIndex, root, paymentCalculationEndTimestamp, activatedAt);
+    }
+
+    /**
+     * @notice Create a RangePayment. Called from both `payForRange` and `payAllForRange`
+     */
+    function _payForRange(RangePayment calldata rangePayment) internal {
+        require(rangePayment.strategiesAndMultipliers.length > 0, "PaymentCoordinator._payForRange: no strategies set");
+        require(rangePayment.amount > 0, "PaymentCoordinator._payForRange: amount cannot be 0");
+        require(
+            rangePayment.duration <= MAX_PAYMENT_DURATION,
+            "PaymentCoordinator._payForRange: duration exceeds MAX_PAYMENT_DURATION"
+        );
+        require(
+            rangePayment.duration % calculationIntervalSeconds == 0,
+            "PaymentCoordinator._payForRange: duration must be a multiple of calculationIntervalSeconds"
+        );
+        require(
+            rangePayment.startTimestamp % calculationIntervalSeconds == 0,
+            "PaymentCoordinator._payForRange: startTimestamp must be a multiple of calculationIntervalSeconds"
+        );
+        require(
+            block.timestamp - MAX_RETROACTIVE_LENGTH <= rangePayment.startTimestamp &&
+                GENESIS_PAYMENT_TIMESTAMP <= rangePayment.startTimestamp,
+            "PaymentCoordinator._payForRange: startTimestamp too far in the past"
+        );
+        require(
+            rangePayment.startTimestamp <= block.timestamp + MAX_FUTURE_LENGTH,
+            "PaymentCoordinator._payForRange: startTimestamp too far in the future"
+        );
+
+        // Require rangePayment is for whitelisted strategy or beaconChainETHStrategy
+        address currAddress = address(0);
+        for (uint256 j = 0; j < rangePayment.strategiesAndMultipliers.length; j++) {
+            IStrategy strategy = rangePayment.strategiesAndMultipliers[j].strategy;
+            require(
+                strategyManager.strategyIsWhitelistedForDeposit(strategy) ||
+                    strategy == delegationManager.beaconChainETHStrategy(),
```

Comment:
Could we just make this an immutable?

Technically this won't do an external call every iteration because of short circuiting, but I see no reason to call in the first place

Reply:
personal preference is that it's not `public` -- but i can live with it!